util/fipstools/acvp/modulewrapper/modulewrapper.cc

```diff
@@ -1205,35 +1216,71 @@ static bool HashMCTSha3(const Span<const uint8_t> args[],
-  // minimum and maximum output length, but the vectors obtained from ACVP
-  // allow only 1024 bits. Supporting dynamically sized MCTs would require
-  // passing the min/max output lengths to the modulewrapper, parsing ibid.,
-  // and dynamically allocating and freeing appropriately sized bufffers. To
-  // keep things simple, we defer that complexity until/if needed.
-  //
-  // https://pages.nist.gov/ACVP/draft-celi-acvp-sha3.html#name-shake-monte-carlo-test
-  const unsigned output_len = 1024/8;
-  const unsigned msg_size = 128/8;
+  const uint8_t* max_outlen_bytes = args[1].data();
+  const uint8_t* min_outlen_bytes = args[2].data();
+  const uint8_t* outlen_bytes = args[3].data();
+
+  // The various output lens are passed to modulewrapper as a length-4 byte array representing
+  // a little-endian unsigned 32-bit integer.
+  uint32_t min_output_len = 0;
+  min_output_len |= min_outlen_bytes[3] << 24;
+  min_output_len |= min_outlen_bytes[2] << 16;
+  min_output_len |= min_outlen_bytes[1] << 8;
+  min_output_len |= min_outlen_bytes[0] << 0;
+
+  uint32_t max_output_len = 0;
+  max_output_len |= max_outlen_bytes[3] << 24;
+  max_output_len |= max_outlen_bytes[2] << 16;
+  max_output_len |= max_outlen_bytes[1] << 8;
+  max_output_len |= max_outlen_bytes[0] << 0;
+
+  uint32_t output_len = 0;
+  output_len |= outlen_bytes[3] << 24;
+  output_len |= outlen_bytes[2] << 16;
+  output_len |= outlen_bytes[1] << 8;
+  output_len |= outlen_bytes[0] << 0;
+
+  uint32_t range = max_output_len - min_output_len + 1;
+
   const size_t array_len = 1001;
-  unsigned char md[array_len][output_len];
-  unsigned char msg[array_len][msg_size];
+  std::vector<std::vector<uint8_t>> md(array_len);
+  std::vector<std::vector<uint8_t>> msg(array_len, std::vector<uint8_t>(16));
 
-  // Zero out |md| and |msg| to clear any residual stack garbage before XOF computation
+  // Zero out |msg| to clear any residual stack garbage before XOF computation
   for (size_t i = 0; i < array_len; i++) {
-    OPENSSL_cleanse(md[i], sizeof(md[0]) * sizeof(unsigned char));
-    OPENSSL_cleanse(msg[i], sizeof(msg[0]) * sizeof(unsigned char));
+    OPENSSL_cleanse(msg[i].data(), 16 * sizeof(uint8_t));
   }
 
-  memcpy(md[0], args[0].data(), msg_size);
+  // The following logic conforms to the SHAKE Monte Carlo tests described in
+  // https://pages.nist.gov/ACVP/draft-celi-acvp-sha3.html#name-monte-carlo-tests-for-sha3-
+  md[0].resize(args[0].size());
+  memcpy(md[0].data(), args[0].data(), args[0].size());
 
-  unsigned output_len_mut = output_len;
   for (size_t i = 1; i < array_len; i++) {
-    memcpy(msg[i], md[i-1], msg_size);
-    EVP_Digest(msg[i], sizeof(msg[i]), md[i], &output_len_mut, MDFunc(), NULL);
+    md[i].resize(output_len);
+
+    if (md[i-1].size() < 16) {
+      memcpy(msg[i].data(), md[i-1].data(), md[i-1].size());
+      size_t pad_size = 16 - md[i-1].size();
+      memset(msg[i].data() + md[i-1].size(), 0, pad_size);
```

Comment:
Isn't everything in msg already zero from line 1251?

Reply:
Yes I believe so. Thanks for pointing that out